streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -208,16 +209,17 @@ public void onCompletion(final RecordMetadata metadata,
             throw new StreamsException(String.format("%sFailed to send record to topic %s due to timeout.", logPrefix, topic));
         } catch (final Exception uncaughtException) {
             throw new StreamsException(
-                String.format(
-                    EXCEPTION_MESSAGE,
-                    logPrefix,
-                    "an error caught",
-                    key,
-                    value,
-                    timestamp,
-                    topic,
-                    uncaughtException.toString()
-                ),
+                    String.format(
+                            EXCEPTION_MESSAGE,
+                            logPrefix,
+                            topic,
+                            uncaughtException.toString(),
+                            "an error caught",
+                            timestamp,
+                            // Key/Value formatted values thrown from MeteredXXX stores.
```

Comment:
This might be ok here. Not 100% sure atm. However, similar to above, this exception does not bubble up into a store (at least no always -- for some case it might bubble into a store -- we need to make sure to handle both cases correctly). Thus, the actual caller of `send()` must handle this case and substitute the actual key/value.

Reply:
I agree. It seems like it will be extremely difficult to trace the code and find all the branches that might throw this exception and verify that they catch it, unwrap it, substitute the right k/v in and re-throw.

Also, it's not just `send()` that can throw this exception, but also `flush` and `close`.

This exception type is in the public API, which indicates to me that it was intended to bubble straight out to users with no intervening catch. This means that we'll start sending "key=%s" instead of "key=[byte array]", which actually seems worse to me.

What do you think about instead just removing k/v from this log message, since we don't always know how to stringify them at this point, and instead doing our best to catch and augment the stack trace at a higher level?

Reply:
Alternatively, I'd recommend creating a checked exception type and adding throws declarations to send, flush, and close, to guarantee that we do in fact catch the exception and augment it.

Reply:
@vvcephei, Thinking in the same line. @mjsax ,  `k/v` in the below log messages : https://github.com/apache/kafka/blob/3c14feb442219f58e0008a6242a24f1a492ca337/streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java#L175
and 
https://github.com/apache/kafka/blob/3c14feb442219f58e0008a6242a24f1a492ca337/streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java#L131
doesn't contain human readable values. Can we remove `key/value` info from the above log messages?

Reply:
Sorry for late reply. I don't think, that `flush()` or `close()` would throw -- during flush/close we if sends go wrong, the corresponding `Callback()` would be triggered and we should throw from `recordSendError()` right?

I would not remove key/value from the log statement. Note, that for regular write into repartition and output topics, both `key` and `value` are actual object and human readable.

Only if `RecordCollector.send()` is called to write into a changelog topic, the data is serialized outside. Thus, it might actually be b cleaner fix, to not use the regular record collector when writing to change log streams and address the issue there. Compare `StoreChangeLogger#logChange` -- we might want to log error messages there?